vortex-sampling-compressor/src/downscale.rs

```diff
@@ -0,0 +1,98 @@
+use vortex_array::array::{PrimitiveArray, PrimitiveEncoding};
+use vortex_array::compute::try_cast;
+use vortex_array::encoding::EncodingVTable;
+use vortex_array::stats::{ArrayStatistics, Stat};
+use vortex_array::{ArrayDType, ArrayData, IntoArrayData, IntoArrayVariant};
+use vortex_dtype::{DType, PType};
+use vortex_error::{vortex_err, VortexResult};
+
+/// Downscale a primitive array to the narrowest PType that fits all the values.
+pub fn downscale_integer_array(array: ArrayData) -> VortexResult<ArrayData> {
+    if !array.is_encoding(PrimitiveEncoding.id()) {
+        // This can happen if e.g. the array is ConstantArray.
+        return Ok(array);
+    }
+    let array = PrimitiveArray::try_from(array)?;
+
+    let min = array
+        .statistics()
+        .compute(Stat::Min)
+        .ok_or_else(|| vortex_err!("Failed to compute min on primitive array"))?;
+    let max = array
+        .statistics()
+        .compute(Stat::Max)
+        .ok_or_else(|| vortex_err!("Failed to compute max on primitive array"))?;
+
+    // If we can't cast to i64, then leave the array as its original type.
+    // It's too big to downcast anyway.
+    let Ok(min) = min.cast(&DType::Primitive(PType::I64, array.dtype().nullability())) else {
+        return Ok(array.into_array());
+    };
+    let Ok(max) = max.cast(&DType::Primitive(PType::I64, array.dtype().nullability())) else {
+        return Ok(array.into_array());
+    };
+
+    downscale_primitive_integer_array(array, i64::try_from(min)?, i64::try_from(max)?)
+        .map(|a| a.into_array())
+}
+
+/// Downscale a primitive array to the narrowest PType that fits all the values.
+fn downscale_primitive_integer_array(
+    array: PrimitiveArray,
+    min: i64,
+    max: i64,
+) -> VortexResult<PrimitiveArray> {
+    if min < 0 || max < 0 {
```

Comment:
Even if the min is > 0 but the ptype is signed you shouldn't convert it to unsigned

Reply:
This is for indices, the caller should know what they're doing